### PR TITLE
EXRExporter: introduces EXR exporter, initial draft.

### PR DESCRIPTION
**Pull request by sciecode**
Initial draft of OpenEXR file format exporter.

Relies internally on WebGL `gl.readPixels` for retrieving render target pixel data information. So it is subject to all of its limitations as well, but it does offer a good support for regular use-cases by default. 

OpenEXR expects the pixel data in linear representation, so we are required to apply EOTF ( Electro-Optical Transfer Function ) before exporting the pixel data. It is also in the best interest to decode encoded data ( e.g RGBEEncoding ), currently the supported texture encodings are mentioned below, If requested, I will expand the encoding support. 

```
- LinearEncoding
- sRGBEncoding
- GammaEncoding
- RGBEEncoding
- RGBM7Encoding
- RGBM16Encoding
```

[EXRExporter Documentation](https://rawcdn.githack.com/sciecode/three.js/eb44934ea943f81d50a48cc6864cfe1ce950bd8e/docs/#examples/en/exporters/EXRExporter)
[EXRExporter Example](https://rawcdn.githack.com/sciecode/three.js/087bea57ce93a8b00aa25883b93c48f0f8070a44/examples/webgl_materials_envmaps_hdr.html)

**Observation**: In working on this pull request I noticed a small inconsistency with `PMREMGenerator`, when generating a PMREM from a LDR input texture, the result RGB channels will be correct, however the alpha channel will be set to 0 for all pixels. This bears no effect when the output PMREM is used as an envmap, but once you export this to an .exr file, depending on the viewer you might just see an entirely black/transparent image. I'll be addressing this issue in a separated PR. It is a very simple fix.

### CODE REVIEW

**Comment by sciecode** [comment]
The current state of this PR does not include an `EXRExporter` example. 

I would like to include one in the final PR, something like exporting PMREMGenerator output or rendered scene HDR snapshot. I'm open to suggestions regarding which workflows to highlight 😄

**Comment by mrdoob** [comment]
No need to do the `examples/js/` version for new files 👌

**Comment by sciecode** [comment]
> No need to do the `examples/js/` version for new files 👌

Oh, I keep forgetting that. Will remove `examples/js` version and modularize entry 👍

**Comment by lokiiarora** [comment]
Hey @sciecode , is this good to go?
I'd love to include this in an internal tool we are building at [Scapic](https://scapic.com) .
Noticed you mentioned something about the PMREM Generator inconsistency, is it something that will get affected if we are using `.fromScene` method.

**Comment by sciecode** [comment]
@lokiiarora this is pretty much good to go.

I've been running test and haven't found any problems related with utilizing the exporter, I'll include a proper example between today/tomorrow, but no changes will be made to the exporter. So if you want to incorporate the current state of the exporter in your current workflow, feel free to do so.

The PMREM inconsistency has already been handled on #19765 and it should be working properly with all methods.

**Comment by mrdoob** [comment]
@sciecode Should we merge this? The PR is still marked as draft.

File: examples/jsm/exporters/EXRExporter.js [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+
+	const exp = a * maxValue;
+	dec.r = r * exp;
+	dec.g = g * exp;
+	dec.b = b * exp;
+
+}
+
+// http://gamedev.stackexchange.com/questions/17326/conversion-of-a-number-from-single-precision-floating-point-representation-to-a/17410#17410
+function encodeFloat16( val ) {
```
Comment (Mugen87) [inline examples/jsm/exporters/EXRExporter.js:538]: Please use [DataUtils.toHalfFloat()](https://threejs.org/docs/index.html#api/en/extras/DataUtils.toHalfFloat) here.

Comment (sciecode) [reply examples/jsm/exporters/EXRExporter.js:538]: Will do.

**Comment by sciecode** [comment]
> @sciecode Should we merge this? The PR is still marked as draft.

There are still some adjustments I would like to make, and I should probably update the Deflate usage as discussed on https://github.com/mrdoob/three.js/issues/20941#issuecomment-753239249.

I'll do my best to work on it next weekend, sorry for the long delay.

**Comment by WestLangley** [comment]
/ping @DavidPeicho You can use this to export PMREMs in .exr format.

**Comment by elalish** [comment]
@sciecode This would be a really nice feature since envMaps are so hard to edit. Not only could you create them from scenes, but you could also do simple things like downsample them based on the minimum roughness in your scene. A nice follow-on would be HDR export. Are you still working on this?

**Comment by sciecode** [comment]
Yes, I've been getting a few requests to resume development on this exporter. Sorry it took this long, wasn't my intention.

I'll focus on rebasing the commit and making sure the exporter is properly working on both Linear and sRGB input render targets, just so we can get a initial version of the exporter live. I should have PR ready by the weekend.

**Pull request closed without merging**